x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/RuleQueryBuilder.java

```diff
@@ -0,0 +1,326 @@
+                                        }
+                                    }
+                                }
+                            }
+                        } else {
+                            logger.warn("Skipping unsupported query rule type [" + rule.type() + "]");
+                        }
+                    }
+                    idSetOnce.set(pinnedIds.stream().distinct().toList());
+                    docsSetOnce.set(pinnedDocs.stream().distinct().toList());
+                    listener.onResponse(null);
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    // Ruleset not found, no rules to apply
+                    idSetOnce.set(Collections.emptyList());
+                    docsSetOnce.set(Collections.emptyList());
+                    listener.onResponse(null);
+                }
+            });
+        });
+
+        QueryBuilder newOrganicQuery = organicQuery.rewrite(queryRewriteContext);
+        RuleQueryBuilder rewritten = new RuleQueryBuilder(
+            newOrganicQuery,
+            matchCriteria,
+            rulesetIds,
+            null,
+            null,
+            idSetOnce::get,
+            docsSetOnce::get
+        );
+        rewritten.boost(this.boost);
+        return rewritten;
+    }
+
+    @Override
+    protected boolean doEquals(RuleQueryBuilder other) {
+        if (this == other) return true;
+        if (other == null || getClass() != other.getClass()) return false;
+        return Objects.equals(rulesetIds, other.rulesetIds)
+            && Objects.equals(matchCriteria, other.matchCriteria)
+            && Objects.equals(organicQuery, other.organicQuery)
+            && Objects.equals(curatedIds, other.curatedIds)
+            && Objects.equals(curatedDocs, other.curatedDocs)
+            && Objects.equals(curatedIdSupplier, other.curatedIdSupplier)
+            && Objects.equals(curatedDocsSupplier, other.curatedDocsSupplier);
+    }
+
+    @Override
+    protected int doHashCode() {
+        return Objects.hash(rulesetIds, matchCriteria, organicQuery, curatedIds, curatedDocs, curatedIdSupplier, curatedDocsSupplier);
+    }
+
+    private static final ConstructingObjectParser<RuleQueryBuilder, Void> PARSER = new ConstructingObjectParser<>(NAME, a -> {
+        QueryBuilder organicQuery = (QueryBuilder) a[0];
+        @SuppressWarnings("unchecked")
+        Map<String, Object> matchCriteria = (Map<String, Object>) a[1];
+        @SuppressWarnings("unchecked")
+        List<String> rulesetIds = (List<String>) a[2];
+        return new RuleQueryBuilder(organicQuery, matchCriteria, rulesetIds);
+    });
+    static {
+        PARSER.declareObject(constructorArg(), (p, c) -> parseInnerQueryBuilder(p), ORGANIC_QUERY_FIELD);
+        PARSER.declareObject(optionalConstructorArg(), (p, c) -> p.map(), MATCH_CRITERIA_FIELD);
```

Comment:
The map is not validated, what objects are we expecting here? A query builder? The test `RuleQueryBuilderTests. testUnknownObjectException` is failing because adding random fields in this map doesn't throw any error.

Reply:
Thanks - I'll make this more strict.